nexus_keycode/protocols/channel_origin_commands.py

```diff
@@ -0,0 +1,458 @@
+        self.auth = auth
+
+    def __str__(self):
+        return self.to_digits()
+
+    def __repr__(self):
+        return (
+            "{}.{}("
+            "{type_code!r}, "
+            "{body!r}, "
+            "{auth!r},))").format(
+            self.__class__.__module__,
+            self.__class__.__name__,
+            **self.__dict__)
+
+    def to_digits(self, obscured=True):
+        # type: (bool) -> str
+        # String of digits making up this Nexus Channel "Token".
+
+        result = "{}{}{}".format(
+            self.type_code,
+            self.body,
+            self.auth
+        )
+        if obscured:
+            # obscure all digits except MAC/auth
+            obscured_digit_count = len(result) - len(str(self.auth))
+            result = full_obscure(result, obscured_digit_count=obscured_digit_count)
+
+        return result
+
+    @staticmethod
+    def digits_from_siphash(siphash_function, digits=6):
+        """ Return the least-significant digits from a Siphash function.
+
+        Defaults to 6, may be increased.
+        """
+        format_str = "{{:0{}d}}".format(digits)
+        return format_str.format(
+            siphash_function.hash() & 0xffffffff
+        )[-digits:]
+
+
+class GenericControllerActionToken(ChannelOriginCommandToken):
+    """ Not intended to be instantiated directly.
+
+    see: `unlink_all_accessories`.
+    """
+
+    _origin_command_type = OriginCommandType.GENERIC_CONTROLLER_ACTION
+
+    def __init__(
+            self,
+            type_,
+            controller_command,
+            auth
+    ):
+        super(GenericControllerActionToken, self).__init__(
+            type_=self._origin_command_type,
+            body=controller_command,
+            auth=self.digits_from_siphash(auth)
+        )
+
+    @enum.unique
+    class GenericControllerActionType(enum.Enum):
+        """ Types of 'generic controller actions' that are possible.
+        Types 0-20 are reserved for Angaza use. Other types may be 'custom'
+        as needed.
+        """
+
+        # Delete all accessories from the receiving controller
+        UNLINK_ALL_ACCESSORIES = 0
+        # Unlock all accessories linked to the receiving controller
+        UNLOCK_ALL_ACCESSORIES = 1
+        # Types 2-99 undefined
+
+    @classmethod
+    def _generic_controller_action_builder(
+        cls,
+        type_,
+        controller_command_count,
+        controller_sym_key
+    ):
+        # type: (int, str) -> ChannelOriginCommandToken
+        """ Resulting token:
+
+        1-digit Origin Keycode Type ID (0)
+        2-digit "Origin Controller Commmand" (0-99)
+        6-digit target authentication (auth for controller)
+        """
+        # Requires 16-byte symmetric Nexus keys
+        assert len(controller_sym_key) == 16
+
+        controller_command_value = type_.value
+
+        packed_target_inputs = bitstring.pack(
+            [
+                "uintle:32=controller_command_count",
```

Comment:
What do you think about adding controller/accessory command counts to the model? I think this would make it much easier to troubleshoot issues with message IDs being entered out of order.

Imagine for example rehydrating a bunch of origin commands and not knowing what the command counts are.

Reply:
I am having trouble understanding the use case - you cannot go backwards and derive the command count from an already-generated origin command, and if you're generating an origin command, you have the command count alongside the generated code. Can you help me understand what you mean by 'rehydrating a bunch of origin commands'?

Reply:
I'm imagining scenarios where a device has had many origin commands generated and we get a question like, "I've entered tokens B (count: 50) and C (count: 51), now why isn't origin command A (count: 5) working?" The reality could be that the commands were delivered out of order or dropped. I think that storing the count would be helpful here because as you point out we can't back-derive it and AFAICT we currently don't store it after we create the origin command and send it off.

We can probably assume for now that the most recently generated origin command has the highest command count, but as things get more complex I think this is risky. So it seems like it would be helpful for a `ChannelOriginCommandToken` to have `self.controller_count` and `self.accessory_count` members to store that context for a backend/database to reference in the future. We do this now for activation tokens too.